pkg/provisioningv2/rke2/planner/planner.go

```diff
@@ -438,70 +510,89 @@ func (p *Planner) reconcile(controlPlane *rkev1.RKEControlPlane, tokensSecret pl
 				if !isUnavailable(entry) {
 					unavailable++
 				}
-				if ok, err := p.drain(entry.Plan.AppliedPlan, plan, entry, clusterPlan, drainOptions); err != nil {
+				if ok, err := p.drain(entry.Plan.AppliedPlan, plan, entry, clusterPlan, drainOptions); !ok && err != nil {
 					return err
-				} else if ok {
-					if err := p.store.UpdatePlan(entry, plan, 0); err != nil {
+				} else if ok && err == nil {
+					// Drain is done (or didn't need to be done) and there are no errors, so the plan should be updated to enact the reason the node was drained.
+					if err = p.store.UpdatePlan(entry, plan, 0); err != nil {
 						return err
+					} else if entry.Metadata.Annotations[rke2.DrainDoneAnnotation] != "" {
+						messages[entry.Machine.Name] = append(messages[entry.Machine.Name], "drain completed")
+					} else if planStatusMessage == "" {
+						messages[entry.Machine.Name] = append(messages[entry.Machine.Name], "waiting for plan to be applied")
 					}
 				} else {
+					// In this case, it is true that ((ok == true && err != nil) || (ok == false && err == nil))
+					// The first case indicates that there is an error trying to drain the node.
+					// The second case indicates that the node is draining.
 					draining = append(draining, entry.Machine.Name)
+					if err != nil {
+						messages[entry.Machine.Name] = append(messages[entry.Machine.Name], err.Error())
+					} else {
+						messages[entry.Machine.Name] = append(messages[entry.Machine.Name], "draining node")
+					}
 				}
 			}
-		} else if !entry.Plan.InSync || (entry.Machine.Status.NodeInfo != nil && entry.Machine.Status.NodeInfo.KubeletVersion != controlPlane.Spec.KubernetesVersion) {
+		} else if planStatusMessage != "" {
 			outOfSync = append(outOfSync, entry.Machine.Name)
-		} else {
-			if ok, err := p.undrain(entry); err != nil {
-				return err
-			} else if !ok {
-				uncordoned = append(uncordoned, entry.Machine.Name)
+		} else if ok, err := p.undrain(entry); !ok && err != nil {
+			return err
+		} else if !ok || err != nil {
+			// The uncordoning is happening or there was an error.
+			// Either way, the planner should wait for the result and display the message on the machine.
+			uncordoned = append(uncordoned, entry.Machine.Name)
+			if err != nil {
+				messages[entry.Machine.Name] = append(messages[entry.Machine.Name], err.Error())
+			} else {
+				messages[entry.Machine.Name] = append(messages[entry.Machine.Name], "waiting for uncordon to finish")
 			}
+		} else if entry.Machine.Status.NodeInfo != nil && !strings.HasPrefix(controlPlane.Spec.KubernetesVersion, entry.Machine.Status.NodeInfo.KubeletVersion) {
+			outOfSync = append(outOfSync, entry.Machine.Name)
+			messages[entry.Machine.Name] = append(messages[entry.Machine.Name], "waiting for kubelet to update")
+		} else {
+			ready = append(ready, entry.Machine.Name)
 		}
 	}
 
-	if required && len(entries) == 0 {
+	if len(entries) == 0 && tierName != "worker" {
```

Comment:
this change removes flexibility from the framework.... I suppose we can re-add it if we need it at a later time.

Reply:
You're right, but I was just trying to simplify the signature of `reconcile`. I can switch it back if you'd like.

Reply:
I would either prefer we use a const for the `tierName` or switch it back to where it was before

Reply:
I will switch it back.